File: internal/models/status_page_settings.go
package models

import "time"

const (
	StatusPageSettingsKey = "status_page"
)

type StatusPageHeadSettings struct {
	Title       string            `bson:"title" json:"title"`
	Description string            `bson:"description" json:"description"`
	Keywords    string            `bson:"keywords" json:"keywords"`
	FaviconURL  string            `bson:"faviconUrl" json:"faviconUrl"`
	MetaTags    map[string]string `bson:"metaTags,omitempty" json:"metaTags,omitempty"`
}

type StatusPageBrandingSettings struct {
	SiteName           string `bson:"siteName" json:"siteName"`
	LogoURL            string `bson:"logoUrl" json:"logoUrl"`
	BackgroundImageURL string `bson:"backgroundImageUrl" json:"backgroundImageUrl"`
	HeroImageURL       string `bson:"heroImageUrl" json:"heroImageUrl"`
}

type StatusPageThemePalette struct {
	PrimaryColor    string `bson:"primaryColor" json:"primaryColor"`
	BackgroundColor string `bson:"backgroundColor" json:"backgroundColor"`
	TextColor       string `bson:"textColor" json:"textColor"`
	AccentColor     string `bson:"accentColor" json:"accentColor"`
}

type StatusPageThemeTypography struct {
	FontFamily string `bson:"fontFamily" json:"fontFamily"`
	FontScale  string `bson:"fontScale" json:"fontScale"`
}

type StatusPageThemeSettings struct {
	Preset     string                    `bson:"preset" json:"preset"`
	Mode       string                    `bson:"mode" json:"mode"`
	Light      StatusPageThemePalette    `bson:"light" json:"light"`
	Dark       StatusPageThemePalette    `bson:"dark" json:"dark"`
	Typography StatusPageThemeTypography `bson:"typography" json:"typography"`
}

type StatusPageLayoutSettings struct {
	Variant string `bson:"variant" json:"variant"`
}

type StatusPageFooterSettings struct {
	Text          string `bson:"text" json:"text"`
	ShowPoweredBy bool   `bson:"showPoweredBy" json:"showPoweredBy"`
}

type StatusPageSettings struct {
	Key       string                     `bson:"key" json:"-"`
	Head      StatusPageHeadSettings     `bson:"head" json:"head"`
	Branding  StatusPageBrandingSettings `bson:"branding" json:"branding"`
	Theme     StatusPageThemeSettings    `bson:"theme" json:"theme"`
	Layout    StatusPageLayoutSettings   `bson:"layout" json:"layout"`
	Footer    StatusPageFooterSettings   `bson:"footer" json:"footer"`
	CustomCSS string                     `bson:"customCss" json:"customCss"`
	UpdatedAt time.Time                  `bson:"updatedAt" json:"updatedAt"`
	CreatedAt time.Time                  `bson:"createdAt" json:"createdAt"`
}

func DefaultStatusPageSettings() StatusPageSettings {
	now := time.Now()
	return StatusPageSettings{
		Key: StatusPageSettingsKey,
		Head: StatusPageHeadSettings{
			Title:       "Statora",
			Description: "Live system status and incident updates.",
			Keywords:    "status, uptime, incidents, maintenance",
			FaviconURL:  "/vite.svg",
			MetaTags:    map[string]string{},
		},
		Branding: StatusPageBrandingSettings{
			SiteName:           "Statora",
			LogoURL:            "",
			BackgroundImageURL: "",
			HeroImageURL:       "",
		},
		Theme: StatusPageThemeSettings{
			Preset: "default",
			Mode:   "system",
			Light: StatusPageThemePalette{
				PrimaryColor:    "#16a34a",
				BackgroundColor: "#f9fafb",
				TextColor:       "#111827",
				AccentColor:     "#0ea5e9",
			},
			Dark: StatusPageThemePalette{
				PrimaryColor:    "#22c55e",
				BackgroundColor: "#0b1220",
				TextColor:       "#e5e7eb",
				AccentColor:     "#38bdf8",
			},
			Typography: StatusPageThemeTypography{
				FontFamily: "Inter, system-ui, -apple-system, BlinkMacSystemFont, \"Segoe UI\", sans-serif",
				FontScale:  "md",
			},
		},
		Layout: StatusPageLayoutSettings{
			Variant: "classic",
		},
		Footer: StatusPageFooterSettings{
			Text:          "",
			ShowPoweredBy: true,
		},
		CustomCSS: "",
		UpdatedAt: now,
		CreatedAt: now,
	}
}

File: apps/web/src/lib/statusPageSettings.ts
import type { StatusPageSettings } from '../types'
import { loadThemePresetStylesheet, getThemePresets, DEFAULT_THEME_PRESET } from './themePresetLoader'

export const STATUS_PAGE_SETTINGS_CACHE_KEY = 'Statora:status-page-settings'

export const DEFAULT_STATUS_PAGE_SETTINGS: StatusPageSettings = {
  head: {
    title: 'Statora',
    description: 'Live system status and incident updates.',
    keywords: 'status, uptime, incidents, maintenance',
    faviconUrl: '/vite.svg',
    metaTags: {},
  },
  branding: {
    siteName: 'Statora',
    logoUrl: '',
    backgroundImageUrl: '',
    heroImageUrl: '',
  },
  theme: {
    preset: DEFAULT_THEME_PRESET,
  },
  footer: {
    text: '',
    showPoweredBy: true,
  },
  customCss: '',
  updatedAt: '',
  createdAt: '',
}

let bootstrappedStatusPageSettings: StatusPageSettings | null = null

export function normalizeStatusPageSettings(settings?: Partial<StatusPageSettings> | null): StatusPageSettings {
  if (!settings) {
    return DEFAULT_STATUS_PAGE_SETTINGS
  }

  const preset = settings.theme?.preset?.trim() || DEFAULT_THEME_PRESET
  const normalizedPreset = preset.endsWith('.css') ? preset : `${preset}.css`

  return {
    head: {
      title: settings.head?.title ?? DEFAULT_STATUS_PAGE_SETTINGS.head.title,
      description: settings.head?.description ?? DEFAULT_STATUS_PAGE_SETTINGS.head.description,
      keywords: settings.head?.keywords ?? DEFAULT_STATUS_PAGE_SETTINGS.head.keywords,
      faviconUrl: settings.head?.faviconUrl ?? DEFAULT_STATUS_PAGE_SETTINGS.head.faviconUrl,
      metaTags: settings.head?.metaTags || {},
    },
    branding: {
      siteName: settings.branding?.siteName ?? DEFAULT_STATUS_PAGE_SETTINGS.branding.siteName,
      logoUrl: settings.branding?.logoUrl ?? '',
      backgroundImageUrl: settings.branding?.backgroundImageUrl ?? '',
      heroImageUrl: settings.branding?.heroImageUrl ?? '',
    },
    theme: {
      preset: normalizedPreset,
      appliedPreset: settings.theme?.appliedPreset,
      mode: settings.theme?.mode,
      overrides: settings.theme?.overrides,
      resolved: settings.theme?.resolved,
    },
    layout: settings.layout,
    footer: {
      text: settings.footer?.text ?? '',
      showPoweredBy: settings.footer?.showPoweredBy ?? true,
    },
    customCss: settings.customCss ?? '',
    updatedAt: settings.updatedAt ?? '',
    createdAt: settings.createdAt ?? '',
  }
}

function upsertMetaTag(selector: string, content: string) {
  const existing = document.head.querySelector(`meta[${selector}]`)
  if (content) {
    if (existing) {
      existing.setAttribute('content', content)
      return
    }

    const meta = document.createElement('meta')
    const [attr, value] = selector.split('=')
    meta.setAttribute(attr, value.replace(/"/g, ''))
    meta.setAttribute('content', content)
    document.head.appendChild(meta)
    return
  }

  if (existing) {
    existing.remove()
  }
}

function setCustomMetaTags(metaTags: Record<string, string>) {
  const existing = document.head.querySelectorAll('meta[data-status-page-meta="true"]')
  existing.forEach(node => node.remove())

  Object.entries(metaTags).forEach(([key, value]) => {
    if (!key || !value) {
      return
    }

    const meta = document.createElement('meta')
    if (key.startsWith('og:') || key.startsWith('twitter:')) {
      meta.setAttribute('property', key)
    } else {
      meta.setAttribute('name', key)
    }
    meta.setAttribute('content', value)
    meta.setAttribute('data-status-page-meta', 'true')
    document.head.appendChild(meta)
  })
}

function upsertFavicon(url: string) {
  let link = document.head.querySelector<HTMLLinkElement>('link[rel="icon"]')
  if (!link) {
    link = document.createElement('link')
    link.rel = 'icon'
    document.head.appendChild(link)
  }
  link.href = url
}

function upsertCustomCss(css: string) {
  const id = 'status-page-custom-css'
  let styleEl = document.getElementById(id) as HTMLStyleElement | null
  if (!styleEl) {
    styleEl = document.createElement('style')
    styleEl.id = id
    document.head.appendChild(styleEl)
  }
  styleEl.textContent = css
}

export function applyStatusPageDocumentSettings(settings: StatusPageSettings): void {
  if (typeof document === 'undefined') {
    return
  }

  document.title = settings.head.title
  upsertMetaTag('name="description"', settings.head.description)
  upsertMetaTag('name="keywords"', settings.head.keywords)
  setCustomMetaTags(settings.head.metaTags)
  upsertFavicon(settings.head.faviconUrl)
  upsertCustomCss(settings.customCss)
}

export function applyStatusPageThemePreset(settings: StatusPageSettings): void {
  const normalized = normalizeStatusPageSettings(settings)
  const presets = getThemePresets().presets
  loadThemePresetStylesheet(normalized.theme.preset, presets).catch(() => {})
}

export function applyStatusPageHeadSettings(settings: StatusPageSettings): void {
  applyStatusPageDocumentSettings(normalizeStatusPageSettings(settings))
}

export function readCachedStatusPageSettings(): StatusPageSettings | null {
  if (typeof window === 'undefined') {
    return null
  }

  try {
    const cached = window.localStorage.getItem(STATUS_PAGE_SETTINGS_CACHE_KEY)
    if (!cached) {
      return null
    }
    const parsed = JSON.parse(cached) as Partial<StatusPageSettings>
    return normalizeStatusPageSettings(parsed)
  } catch {
    return null
  }
}

export function cacheStatusPageSettings(settings: StatusPageSettings): void {
  if (typeof window === 'undefined') {
    return
  }

  const normalized = normalizeStatusPageSettings(settings)
  bootstrappedStatusPageSettings = normalized

  try {
    window.localStorage.setItem(STATUS_PAGE_SETTINGS_CACHE_KEY, JSON.stringify(normalized))
  } catch {
  }
}

export function getBootstrappedStatusPageSettings(): StatusPageSettings | null {
  return bootstrappedStatusPageSettings
}

export function parseStatusPageSettingsPayload(payload: unknown): StatusPageSettings | null {
  if (!payload || typeof payload !== 'object') {
    return null
  }

  const candidate = payload as Partial<StatusPageSettings>
  if (!candidate.head || !candidate.branding || !candidate.theme || !candidate.footer) {
    return null
  }

  return normalizeStatusPageSettings(candidate)
}

export function preloadCachedStatusPageSettings(): StatusPageSettings | null {
  const cachedSettings = readCachedStatusPageSettings()
  if (!cachedSettings) {
    return null
  }

  bootstrappedStatusPageSettings = cachedSettings
  applyStatusPageDocumentSettings(cachedSettings)
  applyStatusPageThemePreset(cachedSettings)

  return cachedSettings
}
